LB-64, Add TLS support

Shwetajain148 opened the pull request.

mchaudhary commented:
@Shwetajain148  next time if you want to change the formatting please do it in separate PR. It is very hard to review the code.

Shwetajain148 commented:
@mchaudhary @mostlyjason This PR consists of two commits.
1- Fix formatting
2- Add TLS Support to the script to send logs securely

In the first commit, I have fixed the indentation. You can see the tab changes here: [Fix Indentation](https://github.com/loggly/install-script/pull/85/commits/fccda8e61286b2f166b9597a25e258128f2be37f)

In second commit you can compare the actual file changes to support TLS. Please go through link here: [Add TLS Support to the script to send logs securely
](https://github.com/loggly/install-script/pull/85/commits/81b1606e847db8456ef4fd9a70da10371e29b9cb)

mostlyjason commented on Linux Script/configure-linux.sh:
```diff
@@ -468,28 +479,97 @@ removeLogglyConf()
+        echo "DOWNLOADING CERTIFICATE"
+        mkdir -pv /etc/rsyslog.d/keys/ca.d
+        curl -O https://logdog.loggly.com/media/logs-01.loggly.com_sha12.crt
+        sudo cp -Prf logs-01.loggly.com_sha12.crt /etc/rsyslog.d/keys/ca.d/logs-01.loggly.com_sha12.crt
```
DOn't forget to clean up the local directory when done

Shwetajain148 replied:
@mostlyjason  Done.

mostlyjason commented on Linux Script/configure-linux.sh:
```diff
@@ -215,18 +219,23 @@ removeLogglyConf()
   case "$LINUX_DIST_IN_LOWER_CASE" in
   *"ubuntu"* )
   echo "INFO: Operating system is Ubuntu."
+  PKG_MGR="apt-get"
```
What are these for?

Shwetajain148 replied:
@mostlyjason : In the case of Rsyslog version 7, we need to install rsyslog-gnutls package. So I have initialised the package manager variable for each Linux distribution. Please refer: [Line-554](https://github.com/loggly/install-script/pull/85/files#diff-fcb43b5e6da2f59da41dc1549eb0fdf9R554)

mostlyjason replied:
Ok thanks

mostlyjason commented on Linux Script/configure-linux.sh:
```diff
@@ -468,28 +479,97 @@ removeLogglyConf()
           #          -------------------------------------------------------
         # Define the template used for sending logs to Loggly. Do not change this format.
-          \$template LogglyFormat,\"<%pri%>%protocol-version% %timestamp:::date-rfc3339% %HOSTNAME% %app-name% %procid% %msgid% [$2@$3] %msg%\n\"
+          \$template LogglyFormat,\"<%pri%>%protocol-version% %timestamp:::date-rfc3339% %HOSTNAME% %app-name% %procid% %msgid% [$LOGGLY_AUTH_TOKEN@$LOGGLY_DISTRIBUTION_ID tag=\\\"Rsyslog\\\"] %msg%\n\"
```
Let's use the tag from the tls documentation

Shwetajain148 replied:
This tag is in the non-tls configuration. I have used the same tag "RsyslogTLS" for the TLS configurations as mentioned in the tls doc. See Tls configurations here:  [7-Conf](https://github.com/loggly/install-script/pull/85/files#diff-fcb43b5e6da2f59da41dc1549eb0fdf9R493)  [8+Conf](https://github.com/loggly/install-script/pull/85/files#diff-fcb43b5e6da2f59da41dc1549eb0fdf9R517)

mostlyjason commented on Linux Script/configure-linux.sh:
```diff
@@ -81,743 +81,739 @@ LOGGLY_DISTRIBUTION_ID="41058"
-	   exit 1
-	fi
+  #This script needs to be run as root
+  if [[ $EUID -ne 0 ]]; then
```
What was wrong with the formatting of the indents before? I noticed that you are changing it but is there any particular reason?

Shwetajain148 replied:
@mostlyjason Formatting is done by IntelliJ IDE automatically. If you want then I can revert it back to the previous indentation and will only raise PR with actual file changes.

Shwetajain148 commented on Linux Script/configure-linux.sh:
```diff
@@ -70,754 +70,835 @@ LOGGLY_USERNAME=
+        )
+        # Send messages to Loggly over TCP using the template.
+        action(type=\"omfwd\" protocol=\"tcp\" target=\"$LOGS_01_HOST\" port=\"$LOGGLY_SYSLOG_TLS_PORT\" template=\"LogglyFormat\" StreamDriver=\"gtls\" StreamDriverMode=\"1\" StreamDriverAuthMode=\"x509/name\" StreamDriverPermittedPeers=\"*.loggly.com\")
+        "
```
These are the TLS configurations for Rsyslog version 7- and 8+. If user sets up for Non-TLS logging then [inputStr_NO_TLS](https://github.com/loggly/install-script/pull/85/files#diff-fcb43b5e6da2f59da41dc1549eb0fdf9R538) will be installed.

Shwetajain148 commented on Linux Script/configure-linux.sh:
```diff
@@ -70,754 +70,835 @@ LOGGLY_USERNAME=
+            SUPPRESS_PROMPT="true"
+            ;;
+            --insecure )
+            LOGGLY_TLS_SENDING="false"
```
Command line parameter to setup non-tls sending.

psquickitprageet commented on Linux Script/configure-linux.sh:
```diff
@@ -89,735 +90,815 @@ LINUX_ENV_VALIDATED="false"
+        logMsgToConfigSysLog "INFO" "INFO: Initiating Configure Loggly for Linux."
+    
+        if [ "$LINUX_ENV_VALIDATED" = "false" ]; then
+        checkLinuxLogglyCompatibility
```
@Shwetajain148:  This statement should appear inside the IF block. Make the same change at all the places where this kind of indentation problem is present.

Shwetajain148 commented:
@mchaudhary @mostlyjason  Please review the script.

- Added the TLS Support.
- Implemented the port accessibility for port 6514 by default
- If the user adds the --insecure in his command then the accessibility of port 514 will be checked

mostlyjason commented on Linux Script/configure-linux.sh:
```diff
@@ -468,32 +476,97 @@ checkAuthTokenAndWriteContents()
+	curl -O https://logdog.loggly.com/media/logs-01.loggly.com_sha12.crt
+	sudo cp -Prf logs-01.loggly.com_sha12.crt /etc/rsyslog.d/keys/ca.d/logs-01.loggly.com_sha12.crt
+	sudo rm logs-01.loggly.com_sha12.crt
+}
```
@Shwetajain148 sorry I had this review in pending for several days and forgot to submit it

Shwetajain148 commented on Linux Script/configure-linux.sh:
```diff
@@ -468,32 +476,97 @@ checkAuthTokenAndWriteContents()
-"
+	"
+if [ "$RSYSLOG_VERSION_TMP" -le "7" ]; then
+				/bin/bash -c "sudo $PKG_MGR install rsyslog-gnutls -y"
```
@mostlyjason  Added the error message. Please review.

mostlyjason replied:
Will this also pass if the package has been previously installed and they are running the script again?

mchaudhary replied:
Yes that should work. I will let @Shwetajain148  comment and then merge this tomorrow morning.

Shwetajain148 replied:
@mostlyjason In the case of already installed package, when script will try to install it, will get the following output:

"rsyslog-gnutls is already the newest version.
0 upgraded, 0 newly installed, 0 to remove and 47 not upgraded."

And then the script will go for the next steps.

mostlyjason replied:
Okay great

Shwetajain148 commented on Linux Script/configure-linux.sh:
```diff
@@ -468,32 +476,112 @@ checkAuthTokenAndWriteContents()
+				then
+				logMsgToConfigSysLog "ERROR" "ERROR: The rsyslog-gnutls package was not downloaded. Please download it and then run the script again."
+				exit 1
+				fi
```
@mostlyjason  I could observe that TLS logging requires the rsyslog-gnutls package in case of rsyslog version 8 as well. Otherwise, logging didn't work. 
Also, I have added the error message in case package doesn't install successfully.

mostlyjason replied:
Okay please test on all the various distributions as well

Shwetajain148 commented on Linux Script/configure-linux.sh:
```diff
@@ -468,32 +476,112 @@ checkAuthTokenAndWriteContents()
+	logMsgToConfigSysLog "ERROR" "ERROR: Certificate could not be downloaded."
+	exit 1
+	fi
+}
```
@mostlyjason I have added the error message in the case when the certificate could not be found.

mostlyjason commented:
@mchaudhary I believe she has addressed all my comments so please merge whenever you're ready
